File: pizza.cpp
// pizza.cpp by Jacob Steinebronn

#include <bits/stdc++.h>
using namespace std;
typedef long long int ll;

// Define some constants based on the bounds
#define MAXN 10
#define MAXK 10
#define MAXMASK 2048
#define MAXUNIQUE 2048
#define MAXM 50000

// Count number of 1 bits in n using Brian Kernighan's Algo
int countSetBits(int n){
    int ans = 0;
    while (n) {
        n &= (n - 1);
        ans++;
    }
    return ans;
}

int n, m, k;
// We will precomp countSetBits for all inputs we care about
int numones[MAXMASK+10];
// dp is our memo table
int dp[MAXK+1][MAXMASK+1][MAXUNIQUE+1];
// This will be for the unique toppings we care about during our dp
vector<int> usables;

int solve(int topsRem, int mask, int idx){
    // If we've used all our toppings or there are none left, 
    // Return the  number of people we've blocked
    if(idx >= usables.size() || topsRem == 0) return numones[mask];
    // Check memoization
    if(dp[topsRem][mask][idx] != -1) return dp[topsRem][mask][idx];

    // Try taking this topping
    int take = solve(topsRem-1, mask | usables[idx], idx+1);
    // Try not taking this topping
    int leave = solve(topsRem, mask, idx+1);

    // Memoize & return with the better of the two
    return dp[topsRem][mask][idx] = max(take, leave);
}

int main(){
    // Precomp num 1 bits for each number we care about
    // to save some time
    for(int i = 0; i < MAXMASK+10; i++){
        numones[i] = countSetBits(i);
    }

    int NUM_TESTS;
    cin >> NUM_TESTS;
    while(NUM_TESTS--){
        cin >> n >> m >> k;

        // Get the bitmasks per topping
        vector<int> toppings(m+1, 0);
        for(int i = 0; i < n; i++){
            int t;
            cin >> t;
            for(int j = 0; j < t; j++){
                int ti;
                cin >> ti;
                // Topping ti can block person i
                toppings[ti] |= 1<<i;
            }
        }

        // Go set up the dp array with -1s
        for(int i = 0; i < n+1; i++){
            for(int j = 0; j < (1<<n)+1; j++){
                for(int kk = 0; kk < (1<<n)+1; kk++){
                    dp[i][j][kk] = -1;
                }
            }
        }
        
        // The idea here is that only 2^n toppings matter, because any more will have
        // duplicates. We can illustrate this idea by taking into consideration
        // two toppings who are each hated by person 1 and 3. We never need to 
        // have both toppings, because adding the second one gives us no benefit
        // Thus, the number of toppings we care about is the number of bitmasks
        // with n bits, or 2^n. We will collect these bitmasks in usables
        usables = vector<int>(0);
        vector<int> usedMasks(MAXMASK+10, 0);
        for(int i = 0; i < toppings.size(); i++){
            if(!usedMasks[toppings[i]]){
                usedMasks[toppings[i]] = 1;
                usables.emplace_back(toppings[i]);
            }
        }

        // If k >= n, we can always pick one topping per person, so no need
        // to do any work finding a minimal solution
        if(k >= n) cout << 0 << endl;
        else cout << n - solve(k, 0, 0) << endl;
    }

    return 0;
}

File: pizza.py
# Solution for Pizza (HSPT 2021 - On Site)
# Author: Atharva Nagarkar

# read in number of test cases
tests = int(input())

# handy function to count number of on bits in a number's binary form
def bitcount(n): return bin(n).count("1")

for tc in range(1, tests + 1):

    # read in input for bounds
    n, m, k = map(int, input().split())
    
    # k (max number of toppings to choose) should never be more than n
    # this is because if you can choose more than n. This is because
    # if k is greater than or equal to n, you can simply choose one topping per person
    k = min(k, n)

    # for each topping, keep a bitmask of the people who hate this topping
    toppings = [ 0 for i in range(m) ]
    
    # read input for the topping preferences
    for i in range(n):
        line = input().split()
        t = int(line[0])
        for j in range(1, t+1):
            x = int(line[j]) - 1
            toppings[x] |= 1 << i

    # only keep around unique topping preferences
    # this reduces our bound on m from 50,000 to be at most 2^n, 
    # which is 1024 (much smaller than 50,000)
    toppings = list(set(toppings))
    m = len(toppings)

    # make our dp memoization table
    # fill it with -1's to specify that we have yet to evaluate the answer for this state
    dp = [ [ [-1 for i in range(1 << n)] for i in range(k + 1)] for i in range(m + 1) ]

    # the recursive dp function
    # the state we are using is dp[topping index][toppings left to choose][bitmask of people]
    def go(at, rem, mask):
        if at == m: # we are at the end, our answer will be n - people we dissatisfied
            return n - bitcount(mask)
        if dp[at][rem][mask] != -1: # we've calculated the answer for this state already
            return dp[at][rem][mask]
        
        ans = go(at + 1, rem, mask) # first transition is to skip this topping

        if rem > 0: # other transition is to consider this topping if possible
            ans = min(ans, go(at + 1, rem - 1, mask | toppings[at]))
        
        # store the answer and return it
        dp[at][rem][mask] = ans
        return ans

    # run the function and print the answer it returns
    print(go(0, k, 0))